### PR TITLE
Feature/033 update metadata

### PR DESCRIPTION
**Pull request by JiceDessaint**

### CODE REVIEW

File: Sources/Core/ModernAVPlayer.swift [hunk 1]
```diff
@@ -110,6 +111,16 @@ public final class ModernAVPlayer: NSObject, MediaPlayer {
...
         context.loadMedia(media: media, autostart: autostart)
     }
+   
+    /**
+     Replaces the current player metadata with the new metadata
+     This function is dedicated for updating live data, and though, will discard duration and isLive previously set.
+     
+     - parameter metadata: metadata to load
+     */
+    public func updateLive(metadata: PlayerMetadata) {
```
Comment (raphrel) [inline Sources/Core/ModernAVPlayer.swift]: I suggest renaming method to -> updateNowPlayingInfo(metadata: PlayerMetadata)

File: Sources/Core/ModernAVPlayer.swift [hunk 2]
```diff
@@ -110,6 +111,16 @@ public final class ModernAVPlayer: NSObject, MediaPlayer {
         }
         context.loadMedia(media: media, autostart: autostart)
     }
+   
+    /**
```
Comment (raphrel) [inline Sources/Core/ModernAVPlayer.swift]: now, all new line documention start with ///

File: Sources/Core/ModernAVPlayer.swift [hunk 3]
```diff
@@ -110,6 +111,16 @@ public final class ModernAVPlayer: NSObject, MediaPlayer {
...
+     Replaces the current player metadata with the new metadata
```
Comment (raphrel) [inline Sources/Core/ModernAVPlayer.swift]: metedata belong to item, not player right nope ?
so Replaces the current item metadata with the new metadata ?

File: Sources/Core/ModernAVPlayer.swift [hunk 4]
```diff
@@ -110,6 +111,16 @@ public final class ModernAVPlayer: NSObject, MediaPlayer {
...
+     This function is dedicated for updating live data, and though, will discard duration and isLive previously set.
```
Comment (raphrel) [inline Sources/Core/ModernAVPlayer.swift]: This method is dedicated for updating nowPlayingInfo data...

File: Sources/Core/Components/PlayerMetadata.swift [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+import AVFoundation
+
+public protocol PlayerMetadata {
```
Comment (raphrel) [inline Sources/Core/Components/PlayerMetadata.swift]: renaming to PlayerMediaMetadata ?

File: Sources/Core/Components/PlayerMetadata.swift [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+public struct ModernAVPlayerMetadata: PlayerMetadata {
+    public let title: String?
+    public let albumTitle: String?
+    public let artist: String?
+    public let localPlaceHolderImageName: String?
+    public let remoteImageUrl: URL?
+    
+    public init(title: String? = nil, albumTitle: String? = nil, artist: String? = nil, localImageName: String? = nil,
```
Comment (raphrel) [inline Sources/Core/Components/PlayerMetadata.swift]: CRLF after each param

File: Sources/Core/Services/NowPlayingService.swift [hunk 7]
```diff
@@ -44,11 +44,14 @@ final class ModernAVPlayerNowPlayingService: NowPlaying {
     }
     private var task: URLSessionTask?
 
-    func update(media: PlayerMedia, duration: Double?) {
-        infos = parseInfos(media: media, duration: duration)
+    func update(metadata: PlayerMetadata?, duration: Double?, isLive: Bool) {
+        infos = parseInfos(metadata: metadata, duration: duration)
+        if #available(iOS 10, *) {
```
Comment (raphrel) [inline Sources/Core/Services/NowPlayingService.swift]: why move this execution here ? not a good place in parseInfos() method ?

File: Example/ViewController.swift [hunk 8]
```diff
@@ -81,6 +81,15 @@ final class ViewController: UIViewController {
     @IBAction func stop(_ sender: UIButton) {
         player.stop()
     }
+    
+    @IBAction func metadata(_ sender: Any) {
+        let newMetadata = ModernAVPlayerMetadata(title: "Updated title",
```
Comment (raphrel) [inline Example/ViewController.swift]: can you add like short timestamp in one of the parameter just to display new fresh data when tap button

File: Sources/Core/ModernAVPlayer.swift (hunk 1 above)
Comment (JiceDessaint) [reply Sources/Core/ModernAVPlayer.swift]: ok

File: Sources/Core/ModernAVPlayer.swift (hunk 2 above)
Comment (JiceDessaint) [reply Sources/Core/ModernAVPlayer.swift]: ok

File: Sources/Core/ModernAVPlayer.swift (hunk 3 above)
Comment (JiceDessaint) [reply Sources/Core/ModernAVPlayer.swift]: yes

File: Sources/Core/ModernAVPlayer.swift (hunk 4 above)
Comment (JiceDessaint) [reply Sources/Core/ModernAVPlayer.swift]: ok

File: Sources/Core/Components/PlayerMetadata.swift (hunk 5 above)
Comment (JiceDessaint) [reply Sources/Core/Components/PlayerMetadata.swift]: Right

File: Sources/Core/Components/PlayerMetadata.swift (hunk 6 above)
Comment (JiceDessaint) [reply Sources/Core/Components/PlayerMetadata.swift]: ok

File: Sources/Core/Services/NowPlayingService.swift (hunk 7 above)
Comment (JiceDessaint) [reply Sources/Core/Services/NowPlayingService.swift]: ParseInfo now takes metadata input, as he should. 
isLive is a property of the stream, not the metadata.
Wdyt ?

File: Example/ViewController.swift (hunk 8 above)
Comment (JiceDessaint) [reply Example/ViewController.swift]: Good idea

File: Example/ViewController.swift [hunk 9]
```diff
@@ -83,11 +83,14 @@ final class ViewController: UIViewController {
     }
     
     @IBAction func metadata(_ sender: Any) {
-        let newMetadata = ModernAVPlayerMetadata(title: "Updated title",
+        let formatter = DateFormatter()
+        formatter.dateFormat = "HH:mm:ss"
+        let timeStamp = formatter.string(from: Date())
+        let newMetadata = ModernAVPlayerMediaMetadata(title: "Updated title (\(timeStamp))",
```
Comment (raphrel) [inline Example/ViewController.swift]: hard to easily read seconds so I suggest to replace all title string by the timeStamp

File: Sources/Core/Services/NowPlayingService.swift (hunk 7 above)
Comment (raphrel) [reply Sources/Core/Services/NowPlayingService.swift]: parseInfos build a new NowPlayingInfo dictionnary. we may first rename the method.
then isLive is a property of the stream, right but this attribute is used as an attibute of the dictionnary

Wdyt ?

File: Sources/Core/ModernAVPlayer.swift [hunk 10]
```diff
@@ -110,6 +111,16 @@ public final class ModernAVPlayer: NSObject, MediaPlayer {
...
     }
+   
+    ///
+    /// Replaces the current item metadata with the new metadata
+    /// This method is dedicated for updating nowPlayingInfo data, and though, will discard duration and isLive previously set.
+    ///
+    /// - parameter metadata: metadata to load
+    ///
+    public func updateNowPlayingInfo(metadata: PlayerMediaMetadata) {
+        context.nowPlaying.update(metadata: metadata, duration: nil, isLive: true)
```
Comment (raphrel) [inline Sources/Core/ModernAVPlayer.swift]: isLive property is media attribute, not metadata like you said.
we may only change metadata by this method, not duration or isLive attributes.

then we can remove "...and though, will discard duration and isLive previously set." from the documentation. 
these attributes will be only set when media is loaded. 

Wdyt ?

File: Example/ViewController.swift [hunk 11]
```diff
@@ -81,6 +81,18 @@ final class ViewController: UIViewController {
     @IBAction func stop(_ sender: UIButton) {
         player.stop()
     }
+    
+    @IBAction func metadata(_ sender: Any) {
+        let formatter = DateFormatter()
+        formatter.dateFormat = "HH:mm:ss"
+        let timeStamp = formatter.string(from: Date())
+        let newMetadata = ModernAVPlayerMediaMetadata(title: "Updated title (\(timeStamp))",
```
Comment (raphrel) [inline Example/ViewController.swift]: hard to easily read all title string in command center (scrolling thing...)
maybe just set timestamp in the title parameter

File: CHANGELOG.md [hunk 12]
```diff
@@ -1,6 +1,10 @@
 # Change Log
 All notable changes to this project will be documented in this file.
 
+## [X.X.X]
+* Feature:
+	* Add function to update metadata in live
```
Comment (raphrel) [inline CHANGELOG.md]: "Add a public method to update now playing info metadata"

File: Sources/Core/ModernAVPlayer.swift (hunk 10 above)
Comment (JiceDessaint) [reply Sources/Core/ModernAVPlayer.swift]: Ok, added a new function update(metadata:)

File: Example/ViewController.swift (hunk 11 above)
Comment (JiceDessaint) [reply Example/ViewController.swift]: done

File: CHANGELOG.md (hunk 12 above)
Comment (JiceDessaint) [reply CHANGELOG.md]: done

File: Sources/Core/Services/NowPlayingService.swift [hunk 13]
```diff
@@ -45,12 +46,20 @@ final class ModernAVPlayerNowPlayingService: NowPlaying {
...
         }
+        if let duration = duration,
+            duration.isNormal {
+            infos[MPMediaItemPropertyPlaybackDuration] = duration
+        }
         MPNowPlayingInfoCenter.default().nowPlayingInfo = infos
     }
+
+    func update(metadata: PlayerMediaMetadata) {
+        MPNowPlayingInfoCenter.default().nowPlayingInfo = parse(metadata: metadata)
```
Comment (raphrel) [inline Sources/Core/Services/NowPlayingService.swift]: instance infos variable should be also set with new metadata.
we have to think again about the parse method

Comment (JiceDessaint) [reply Sources/Core/Services/NowPlayingService.swift]: What about that new commit I just push?

File: Sources/Core/Services/NowPlayingService.swift [hunk 14]
```diff
@@ -39,31 +39,32 @@ protocol NowPlaying {
...
             duration.isNormal {
             infos[MPMediaItemPropertyPlaybackDuration] = duration
         }
-        MPNowPlayingInfoCenter.default().nowPlayingInfo = infos
+        self.infos = infos
     }
 
     func update(metadata: PlayerMediaMetadata) {
-        MPNowPlayingInfoCenter.default().nowPlayingInfo = parse(metadata: metadata)
+        infos = parse(metadata: metadata)
```
Comment (raphrel) [inline Sources/Core/Services/NowPlayingService.swift]: MPNowPlayingInfoPropertyIsLiveStream & MPMediaItemPropertyPlaybackDuration will be erased so not a good solution. parse should update dictionnary, not create a new one.

**Pull request merged**